app/models/template_input.rb

```diff
@@ -95,7 +103,13 @@ def ready?
     end
 
     def resolved_value
-      input_value.value
+      value = input_value.value
+
+      if @input.options_array.blank? || @input.options_array.include?(value)
+        value
+      else
+        raise(ValueNotPermitted, "Value '#{value}' is not a valid selection for #{@input.name}")
```

Comment:
I think this should be handled in `ready?` so it returns false for invalid value, otherwise when you preview a template with invalid input it will fail instead of displaying $INPUT placeholder